Have other agents wait for vault status(/persist/vault), before proceeding

cshari-zededa opened the pull request:
- So far, it is driven through device-steps.sh
- This PR is in a step towards pushing that logic inside agents
- downloader/verifier/volumemgr now wait for vault status before
  starting any work
- vaultmgr is still called in device-steps twice, that is to handle
  upgrade from old images where /persist/img and /persist/config
  may be encrypted i.e. setupDeprecatedVaults will be removed once
  all the devices move to a sentinel release (with this change)
- This PR also sets up the stage, where /persist/vault may only be
  unlocked after attestation is successful

Signed-off-by: Hariharasubramanian C S <cshari@zededa.com>

eriknordmark commented on pkg/pillar/cmd/volumemgr/volumemgr.go:
```diff
@@ -111,11 +114,6 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 	if err := pidfile.CheckAndCreatePidfile(log, agentName); err != nil {
 		log.Fatal(err)
 	}
-	log.Infof("Starting %s", agentName)
```
Why did you remove this line?

cshari-zededa replied:
Corrected it

eriknordmark commented on pkg/pillar/cmd/volumemgr/volumemgr.go:
```diff
@@ -430,6 +441,28 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
+	}
+	log.Infof("processed Vault Status")
+
+	log.Infof("Starting %s", agentName)
```
Started a long time ago ...

cshari-zededa replied:
Corrected it

eriknordmark commented on pkg/pillar/scripts/device-steps.sh:
```diff
@@ -133,8 +133,8 @@ touch "$WATCHDOG_PID/zedbox.pid" "$WATCHDOG_FILE/zedbox.touch"
 #It is a device with TPM, enable disk encryption
-    if ! $BINDIR/vaultmgr setupVaults; then
-        echo "$(date -Ins -u) device-steps: vaultmgr setupVaults failed"
+    if ! $BINDIR/vaultmgr setupDeprecatedVaults; then
```
Where do we setup up the non-deprecated vault?

cshari-zededa replied:
I guess you are referring to /persist/vault handling. That is done when vaultmgr starts with "runAsService". The idea is that, at some point (when we no longer have /persist/config and /persist/img as vaults on any of the EVE devices), we can remove this block of code from device-steps.sh, and just retain the part where vaultmgr is started with runAsService (at that point, we no longer need that argument, we can simply start vaultmgr along with other agents).

eriknordmark reviewed (commented):
Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
 - subscribe
 - handle the change
 - poke StillRunning
 - unsubscribe

Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.

cshari-zededa commented:
> Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
> 
> * subscribe
> * handle the change
> * poke StillRunning
> * unsubscribe
> 
> Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
> But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.
> 
> That might make sense in the utils package; a utils/waitfor.go??

Sure, I can fold the current ones into utils/waitfor.go, and probably we can tackle unsubscribe as an incremental change in a follow-up PR?

cshari-zededa commented:
> Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
> 
> * subscribe
> * handle the change
> * poke StillRunning
> * unsubscribe
> 
> Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
> But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.
> 
> That might make sense in the utils package; a utils/waitfor.go??

I've introduced utils/waitfor.go, and updated the code in verifier/volumemgr/downloader to call this function.
Can you please take a look at the updated changes?

eriknordmark commented on pkg/pillar/utils/waitfor.go:
```diff
@@ -0,0 +1,65 @@
+	ctx := ctxArg.(*Context)
+	vault := statusArg.(types.VaultStatus)
+	if vault.Name == types.DefaultVaultName &&
+		vault.Status != info.DataSecAtRestStatus_DATASEC_AT_REST_ERROR {
```
Interesting question what we might want to do if this fails. I guess zedagent will come up and report it. That's good enough for now.

cshari-zededa commented on pkg/pillar/cmd/vaultmgr/vaultmgr.go:
```diff
@@ -27,26 +28,26 @@ import (
+	pubVaultStatus       pubsub.Publication
+	subGlobalConfig      pubsub.Subscription
+	GCInitialized        bool // GlobalConfig initialized
+	defaultVaultUnlocked bool
```
This is just setting it up for the upcoming PR, where we will receive escrowed data back from zedagent in vaultmgr, and we will try unlocking the default vault, if it is not still unlocked.

eriknordmark replied:
OK

eriknordmark reviewed (approved):
LGTM. Thanks for the utils function. I'll follow that pattern and add other utils.
